cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/StartProgramEventSubscriber.java

```diff
@@ -53,23 +54,30 @@ public class StartProgramEventSubscriber extends EventSubscriber {
   private final CConfiguration cConf;
   private final EventReaderProvider<StartProgramEvent> extensionProvider;
   private final ProgramLifecycleService lifecycleService;
+  private final RunRecordMonitorService runRecordMonitorService;
+  private final int maxConcurrentRuns;
   private ScheduledExecutorService executor;
   private Collection<EventReader<StartProgramEvent>> readers;
   private ExecutorService threadPoolExecutor;
 
   /**
    * Create instance that handles StartProgramEvents.
    *
-   * @param cConf             CDAP configuration
-   * @param extensionProvider eventReaderProvider for StartProgramEvent Readers
-   * @param lifecycleService  to publish start programs to TMS
+   * @param cConf                   CDAP configuration
+   * @param extensionProvider       eventReaderProvider for StartProgramEvent Readers
+   * @param lifecycleService        to publish start programs to TMS
+   * @param runRecordMonitorService basic flow-control
    */
   @Inject
-  StartProgramEventSubscriber(CConfiguration cConf, EventReaderProvider<StartProgramEvent> extensionProvider,
-                              ProgramLifecycleService lifecycleService) {
+  StartProgramEventSubscriber(CConfiguration cConf,
+                              EventReaderProvider<StartProgramEvent> extensionProvider,
+                              ProgramLifecycleService lifecycleService,
+                              RunRecordMonitorService runRecordMonitorService) {
```

Comment:
Is RunRecordMonitorService initialized and available before this class? If there is a chance that it may not be ready, we should have logic to make sure this service does not crash and retries again.

Reply:
Good catch. RunRecordMontorService starts when AppFabricServer starts. I believe it'll be injected, but may not be in starting state. We can have a helper method to check if it's in running state or not, and log WARN if it's not in running state. WDYT?

Reply:
Do we want it to be in RUNNING state or STARTING STATE?